turbopack/crates/turbo-tasks/src/parallel.rs

```diff
@@ -0,0 +1,535 @@
+            index: 0,
+            handle: Handle::current(),
+            turbo_tasks: try_turbo_tasks(),
+            span: Span::current(),
+            phantom: std::marker::PhantomData,
+        }
+    }
+
+    fn task<F>(&mut self, f: F) -> JoinHandle<()>
+    where
+        F: FnOnce() -> R + Send + 'l,
+    {
+        struct SendablePtr<T>(*mut Option<T>);
+        unsafe impl<T: Send> Send for SendablePtr<T> {}
+        unsafe impl<T: Sync> Sync for SendablePtr<T> {}
+        impl<T> SendablePtr<T> {
+            fn new(reference: &mut Option<T>) -> Self {
+                SendablePtr(reference as *mut Option<T>)
+            }
+
+            unsafe fn get_mut(&mut self) -> &mut Option<T> {
+                // SAFETY: This is a valid pointer, as we got this pointer from a reference.
+                unsafe { &mut *self.0 }
+            }
+        }
+
+        let mut result_cell = SendablePtr::new(&mut self.results[self.index]);
+        self.index += 1;
+
+        let f: Box<dyn FnOnce() + Send + 'l> = Box::new(move || {
+            let result = f();
+            // SAFETY: This is a valid pointer, as we got this pointer from a reference.
+            let result_cell = unsafe { result_cell.get_mut() };
+            *result_cell = Some(result);
+        });
+        // SAFETY: In `process_in_parallel` we ensure that the spawned tasks is awaited before the
+        // lifetime `'l` ends.
+        let f: Box<dyn FnOnce() + Send + 'static> = unsafe {
+            transmute::<Box<dyn FnOnce() + Send + 'l>, Box<dyn FnOnce() + Send + 'static>>(f)
+        };
+        let turbo_tasks = self.turbo_tasks.clone();
+        let span = self.span.clone();
+        self.handle.spawn(async move {
+            turbo_tasks_try_scope(turbo_tasks, || {
+                let _guard = span.entered();
+                f();
+            })
+        })
+    }
+
+    /// Converts the context into a vector of results
+    ///
+    /// ## Safety
+    ///
+    /// The caller must ensure that all tasks have been awaited before calling this method.
+    unsafe fn into_results(self) -> Vec<Option<R>> {
+        self.results.into_vec()
+    }
+}
+
+/// Helper method to spawn tasks in parallel, ensuring that all tasks are awaited and errors are
+/// handled. Also ensures turbo tasks and tracing context are maintained across the tasks.
+///
+/// ## Safety
+///
+/// The caller must ensure that all references in `inner` are valid for the lifetime `'l`.
+unsafe fn process_in_parallel<'l, I, R>(len: usize, inner: I) -> Vec<Option<R>>
+where
+    R: Send + 'l,
+    I: FnOnce(&mut ProcessInParallelContext<'l, R>) -> Vec<JoinHandle<()>> + 'l,
+{
+    let mut process_context = ProcessInParallelContext::new(len);
+    block_in_place(|| {
+        let tasks = inner(&mut process_context);
+        process_context.handle.block_on(
+            async {
+                let mut first_err = None;
+                for task in tasks {
+                    match task.await {
```

Comment:
What happens if the tokio runtime shuts down? https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#shutdown

You've asserted that `process_in_parallel` finishes after the spawned tasks do, but it might exit early if the tokio runtime shuts down?

It's pretty famously hard/impossible to get scoped async tasks working safely in Rust:
- https://github.com/tokio-rs/tokio/issues/3162
- https://without.boats/blog/the-scoped-task-trilemma/
- https://github.com/rmanoka/async-scoped/issues/33

This might be doable because you're willing to block the thread (`block_in_place`), but that obviously has some significant performance tradeoffs.

Reply:
Handled this case by using a new runtime. I stole that from tokio_scoped.